File: backend/app/routers/scenarios.py
import logging
from typing import List, Optional
from fastapi import APIRouter, HTTPException

from backend.app.skills.registry import get_domain_skills, list_scenarios_by_skill
from backend.app.skills.direct_path import resolve_params, execute_scenario, format_result
from backend.app.schemas import (
    ScenarioSummary,
    ScenarioParamsResponse,
    ScenarioExecuteRequest,
    ScenarioExecuteResponse,
)

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/api/scenarios", tags=["scenarios"])


def is_direct_path_enabled(scenario: dict) -> bool:
    """判定场景是否开启快捷直通查询能力 (支持显式标志与模板特征判定)"""
    if "direct_path_enabled" in scenario:
        return bool(scenario["direct_path_enabled"])
    return bool(scenario.get("sql_template_refs")) and bool(scenario.get("default_template"))


@router.get("", response_model=List[ScenarioSummary])
def list_scenarios_tree():
    """获取全量业务领域及其下属快捷场景列表 (自动过滤仅 LLM 场景)。"""
    summary_list = []
    domain_skills = get_domain_skills()
    for domain_name, domain_info in domain_skills.items():
        scenarios_items = []
        for s in list_scenarios_by_skill(domain_name):
            if is_direct_path_enabled(s):
                scenarios_items.append({
                    "name": s["name"],
                    "title": s.get("title", s["name"]),
                    "description": s.get("description", ""),
                    "direct_path_enabled": True,
                })
        if scenarios_items:
            summary_list.append({
                "domain": domain_name,
                "domain_title": domain_info.get("title") or domain_name.replace("_", " ").title(),
                "scenarios": scenarios_items,
            })
    return summary_list


@router.get("/{domain}/{scenario}/params", response_model=ScenarioParamsResponse)
def get_scenario_params_endpoint(domain: str, scenario: str, template_name: Optional[str] = None):
    """解析获取指定场景的参数定义与模板元数据。"""
    try:
        data = resolve_params(domain, scenario, template_name=template_name)
        return data
    except ValueError as e:
        raise HTTPException(status_code=404, detail=str(e))
    except Exception as e:
        logger.error("Failed to resolve scenario params for %s/%s: %s", domain, scenario, e)
        raise HTTPException(status_code=500, detail=f"Failed to resolve scenario params: {e}")


@router.post("/{domain}/{scenario}/execute", response_model=ScenarioExecuteResponse)
def execute_scenario_endpoint(domain: str, scenario: str, request: ScenarioExecuteRequest):
    """直通安全执行指定场景的 SQL 查询并返回格式化结果。"""
    try:
        params_info = resolve_params(domain, scenario, template_name=request.template_name)
        output_type = params_info.get("output_type", "table")
        
        req_page = request.page or 1
        req_page_size = request.page_size or 50

        rows, columns, total_count = execute_scenario(
            domain_name=domain,
            scenario_name=scenario,
            params=request.params,
            template_name=request.template_name,
            page=req_page,
            page_size=req_page_size,
        )
        formatted_data = format_result(
            rows=rows,
            columns=columns,
            output_type=output_type,
            total_count=total_count,
            page=req_page,
            page_size=req_page_size,
        )
        return formatted_data
    except ValueError as e:
        raise HTTPException(status_code=404, detail=str(e))
    except Exception as e:
        logger.error("Failed to execute scenario %s/%s: %s", domain, scenario, e)
        raise HTTPException(status_code=500, detail=f"Failed to execute scenario query: {e}")

File: frontend/src/api/scenarios.ts
// frontend/src/api/scenarios.ts
import api from './index'

export interface ScenarioItemSummary {
  name: string
  title: string
  description: string
  direct_path_enabled?: boolean
}

export type ScenarioSummary = ScenarioItemSummary

export interface ScenarioDomainSummary {
  domain: string
  domain_title: string
  scenarios: ScenarioItemSummary[]
}

export interface ParameterOption {
  value: string
  label: string
}

export interface ParameterDef {
  type: string
  widget: string
  description: string
  required: boolean
  default: string | number | null
  options: ParameterOption[]
}

export interface TemplateInfo {
  name: string
  label: string
}

export interface ScenarioParamsMeta {
  name: string
  title: string
  output_type: string
  templates?: TemplateInfo[]
  default_template?: string
  parameters: Record<string, ParameterDef>
}

export interface TableQueryResult {
  type: 'table'
  columns: string[]
  rows: (string | number)[][]
  row_count: number
  total_count?: number
  page?: number
  page_size?: number
  total_pages?: number
  is_truncated?: boolean
}

export interface ScalarQueryResult {
  type: 'scalar'
  value: string | number
  label: string
}

export type ScenarioQueryResult = TableQueryResult | ScalarQueryResult

/**
 * 拉取全量场景领域树列表
 */
export async function getScenariosApi(): Promise<ScenarioDomainSummary[]> {
  const data = await api.get('/api/scenarios')
  return data as unknown as ScenarioDomainSummary[]
}

/**
 * 解析获取指定场景的参数定义与模板
 */
export async function getScenarioParamsApi(
  domain: string,
  scenario: string,
  templateName?: string
): Promise<ScenarioParamsMeta> {
  const params: Record<string, string> = {}
  if (templateName) {
    params.template_name = templateName
  }
  const data = await api.get(`/api/scenarios/${domain}/${scenario}/params`, { params })
  return data as unknown as ScenarioParamsMeta
}

/**
 * 执行快捷场景直通 SQL 查询
 * 2026-08-30: 直通查询为真实 SQL 执行，单独放宽超时至 60s，避免慢查询被全局 10s 超时整单中断
 */
export async function executeScenarioApi(
  domain: string,
  scenario: string,
  userParams: Record<string, any>,
  templateName?: string,
  page: number = 1,
  pageSize: number = 50
): Promise<ScenarioQueryResult> {
  const payload = {
    params: userParams,
    template_name: templateName,
    page,
    page_size: pageSize,
  }
  const data = await api.post(`/api/scenarios/${domain}/${scenario}/execute`, payload, {
    timeout: 60000,
  })
  return data as unknown as ScenarioQueryResult
}
